cobra/core/reaction.py

```diff
@@ -263,21 +266,111 @@ def bounds(self, value):
 
     @property
     def flux(self):
-        """Reaction flux in the most recent solution."""
-        if self._model is None:
-            raise RuntimeError("not part of a model")
-        if self._model.solution is None:
-            raise RuntimeError("model has not been solved")
-        return self._model.solution[self.id]
+        """
+        The flux value in the most recent solution.
+
+        Flux is the primal value of the corresponding variable in the model.
+
+        Warnings
+        --------
+        * Accessing reaction fluxes through a `Solution` object is the safer,
+          preferred, and only guaranteed to be correct way. You can see how to
+          do so easily in the examples.
+        * Reaction flux is retrieved from the currently defined
+          `self._model.solver`. The solver status is checked but there are no
+          guarantees that the current solver state is the one you are looking
+          for.
+        * If you modify the underlying model after an optimization, you will
+          retrieve the old optimization values.
+
+        Raises
+        ------
+        RuntimeError
+            If the underlying model was never optimized beforehand or the
+            reaction is not part of a model.
+        OptimizationError
+            If the solver status is anything other than 'optimal'.
+        AssertionError
+            If the flux value is not within the bounds.
+
+        Examples
+        --------
+        >>> import cobra
+        >>> import cobra.test
+        >>> model = cobra.test.create_test_model("textbook")
+        >>> solution = model.optimize()
+        >>> model.reactions.PFK.flux
+        7.477381962160283
+        >>> solution.fluxes.PFK
+        7.4773819621602833
+        """
+        try:
```

Comment:
I agree they are a lot of checks and I would like to reduce them. However, the detailed checks are there in order to provide as helpful feedback as possible. If I just get `None` as a user I will have to hunt down the source myself rather than being pointed at the likely reason right away. Hopefully the first try will simply go through in 99% of the cases.